### PR TITLE
instance spec rework: flatten InstanceSpecV0

### LINKED ISSUE #735: instance specs could be improved
# The problem(s)

The instance spec types in the `propolis_api_types` crate are an important conceptual part of the live migration protocol, but there are several things about the way I implemented them that leave things to be desired:

- **Well-formed specs are not necessarily valid**

It's easy to construct an instance spec that type-checks but can't be used to start an instance. For example, it's trivial to have a spec with a storage device with no valid backend.
- **Specs have a complex hierarchy of component types**

Instance specs are composed of device specs and backend specs, each of which are further divided into their own fields and sub-collections. This makes adding new components while respecting the spec versioning rules something of a hassle (and the extra structure doesn't even guarantee the specs are valid!).
- **Not all components have names**

Live migration assumes that all (migratory) VM components have names. This allows Propolis to send a source component's saved state to the correct counterpart on the target. While many spec components are stored in a `HashMap<String, Component>`, some are not, so Propolis has to generate names for them, which then have to be kept consistent between Propolis versions.

- **There are three different spec builders**

One lives in `propolis_api_types`. `propolis-server` then wraps that in its own builder, which can also convert config TOML entries and `InstanceEnsureRequest` components to spec components. Yet a third builder lives in `propolis-client`; it's basically the same as the builder in `propolis_api_types`, except that it operates on Progenitor-generated types instead of the raw types.

- **`propolis_api_types` contains a lot more than just types**

For a crate that's ostensibly just a bag of types, this crate has a *lot* of code, most of which is only ever used by Propolis servers. Other Propolis components that want Just The Types, Please shouldn't need to pull in all the server-only methods.

- **Some component types are conditionally compiled**

The SoftNPU component definitions in `propolis_api_types` are behind Falcon feature guards. This forces us to maintain two separate OpenAPI documents, one with Falcon enabled and one without, which can lead to annoying rake-steps in CI if one forgets to check the build for one of the two variants. (In fact, as I write this, PHD builds are broken when Falcon is enabled because of a missing `patch` directive in the Falcon client generator...)

# Proposed solutions

I propose the following changes to fix these problems:

- **Create an internal `Spec` type for propolis-server**

This type is the VM component manifest type that Propolis server uses internally (in particular during VM setup). This spec isn't `Serialize`, isn't part of the API, and isn't used in the live migration protocol. Because this spec can be structured however we like without concern for versioning it, it can (and will) be structured so that illegal states are unrepresentable.

- **Flatten `propolis_api_types::v0::InstanceSpecV0`**

Instead of a complex hierarchy of component types, the "on the wire" spec used in the API and LM protocol is just a mainboard and a `HashMap<String, Component>`. Adding a new component type is very easy: just define a struct for it and add it to `enum Component`. Because all components are now in a map, they all have names.

- **Define conversions between the two spec types in propolis-server**

propolis-server implements `From<InternalSpec> for VersionedInstanceSpec>` and `TryFrom<VersionedInstanceSpec> for InternalSpec` and converts to/from its internal type at the API and LM protocol boundaries. The `TryFrom` impl gives us a "parse, don't validate" discipline for incoming API specs.

- **Just one builder in propolis-server**

propolis-server understands how to convert config TOML elements and `InstanceEnsureRequest` fields into spec elements and feeds them into a single `InternalSpec` builder. The `TryFrom<VersionedInstanceSpec>` impl also operates in terms of this builder. The `propolis_api_types` builder and `propolis_client` builders are removed. The only user of the latter (PHD) builds API specs simply by adding components to a new V0 spec's map. (This plan assumes there won't eventually be other Propolis clients who want to initialize VMs using raw specs. I had originally planned for the control plane to do this, but have changed my mind and now think it should continue using the existing instance-ensure API. See the headnote to RFD 344 for more details.)

- **Move migration compat checks into propolis-server**

The migration compatibility traits that live in `propolis_api_types` are only useful in the LM protocol. They should live in propolis-server instead (which can then implement them in terms of its internal spec representation, if it's convenient to do so).

- **Remove `#[cfg(feature)]` guards from API types**

Instead of compiling components out and having different clients built with different feature sets, there's just one client that defines all possible component types. If a server doesn't support a component a client has asked for, it rejects the spec (once again made easier by having an internal spec type to parse into--if a feature is compiled out of the server, then the internal type won't have fields for its components).

---

I have drafted a set of changes that achieves these goals and am pretty happy with how it looks; indeed, I've mostly filed this issue to have a wall of text to link the PRs to. I'll start putting those up for review presently.

Doing this now lays the groundwork for adding CPUID information to instance specs and initializing migration targets from sources' specifications. (It's also much easier to correct foundational mistakes like these while the API and migration protocol can still change without strictly requiring new versions of everything.)

### PR DESCRIPTION
**Pull request by gjcolombo**
(9/9 in the instance spec rework; see #735.)

Finally, the point of all the prior refactoring PRs!

Remove the complex hierarchy of component types from `InstanceSpecV0` and instead turn V0 specs into a `Board` and a bag of `Component`s. This makes it much easier to define new component types flexibly: 

- New components all go into the same collection: no more deciding if you've defined a device or a backend or deciding what to do if you have something in between.
- New components always appear in a map and so always have names (their keys). A bonus is that having only one component map prevents keys from being reused in multiple maps.
- Because new components don't need fields, there's no need to remember to tag component fields with serde's `default` or `skip_serializing_if` attributes (though this may still be needed for component fields).
- It's also much harder to get painted into a backwards-compatibility corner. Suppose you define a new component, add a field for it, and make the field an `Option`. If you later want to support having more than one of that component, you need to add a new spec field with higher maximum cardinality to avoid breaking old specs. Now you have two fields for the same kind of component. Yuck.

The downside of all this is that there are now more wire specs that type check but that the server should reject; for example, you can now have a wire spec with multiple panic devices. The previous changes in this series mitigate this by converting wire specs into more rigorously organized internal specs (which may enforce e.g. cardinality limits) before propolis-server will actually use them.

Now that wire specs have a much simpler structure, the spec builder in the propolis-client lib pulls very little weight, so remove it. Incoming wire specs need to be validated on the server in any case, and two different server versions may disagree on whether a particular wire spec is acceptable (if one has features the other does not), so it's hard to have a single builder that checks invariants for all relevant server versions. Clients who want to have a mistake-catching/invariant-protecting builder are, of course, still free to define their own (and those who might not want one, like PHD, don't need to pull one in that they won't use).

Finally, remove the `#[cfg(feature = "falcon")]` guards from the api-types crate. Servers built without Falcon support will reject specs that contain Falcon components. The downside to this is that users of generated clients who never intend to access a Falcon server (e.g. sled agent) will start to see these types. The upside is that this means that propolis-server's OpenAPI definition no longer varies with its feature set, which means we no longer need `propolis-server-falcon.json` (which is easy to forget to update) or the Falcon variant of the generated client (whose Progenitor directives had to be manually kept in sync with the non-Falcon client).

This is (once again) a migration protocol-breaking change, so the migrate-from-base tests are (once again) expected to fail.

Tests: cargo test, PHD.

Closes #735.

### CODE REVIEW

File: bin/propolis-server/src/lib/spec/api_spec_v0.rs
Comment (hawkw) [inline bin/propolis-server/src/lib/spec/api_spec_v0.rs:188]: i kinda wondered if it was worth having functions like `ComponentV0::is_storage_backend() -> bool` but, upon further consideration, we would only ever call them in one place, so probably not...

**Review by hawkw — APPROVED** [review]
Overall, this looks good to me --- I did try to  search for specific nits to pick and didn't really find any. It's a bit of a shame to do more of the instance-spec validation at runtime, but I agree that it's the right path forward given everything you wrote in the commit message.

**Comment by gjcolombo** [comment]
Thanks for the review, @hawkw! I'm going to hold this until #756 lands to save us from having to spend time rebasing that change onto this one.

**Comment by gjcolombo** [comment]
In addition to failing CI, my resolution of the conflicts from my force-push (to rebase on top of #756) is not as clean as it probably could be. I'll push another commit or two soon to clean things up.

File: crates/propolis-api-types/src/instance_spec/components/devices.rs
Comment (iximeow) [inline crates/propolis-api-types/src/instance_spec/components/devices.rs:108]:
```suggestion
/// how to boot a guest.
```

minorest quibble: the settings here are processed before the guest has landed in an OS for the most part. unless there's a way we'd expect to speak to guests via UEFI (which i genuinely would believe, and would believe if there's something specific to OSes that we anticipate and i dunno about)

File: bin/propolis-server/src/lib/spec/builder.rs
Comment (iximeow) [inline bin/propolis-server/src/lib/spec/builder.rs:148]: fwiw i think this is of substantially less utility since you must have called `add_boot_order()` with at least an empty list for this to not panic on accessing `boot_settings`. as-is, most of this function, by vertical space, is just dancing around the fact that the caller just set `self.spec.boot_settings` to `Some`.

lets fold this into `add_boot_order` and be done with it?

**Review by iximeow — APPROVED** [review]
i forgot how much nicer components were going to be once #756 landed! two minor quibbles about boot settings though :)

File: crates/propolis-api-types/src/instance_spec/components/devices.rs
Comment (gjcolombo) [reply crates/propolis-api-types/src/instance_spec/components/devices.rs]: I took this in a slightly different direction in 04fb943, but I think I've captured the gist of this (this is information for the firmware image to use, not--generally--the next-stage binary that it finds and transfers control to).

Comment (iximeow) [reply crates/propolis-api-types/src/instance_spec/components/devices.rs]: oh yeah i like that a lot, thanks!

**Pull request merged**
